api/v1beta1/types.go

```diff
@@ -520,13 +523,44 @@ type BlockDeviceVolume struct {
 	// +optional
 	Type string `json:"type,omitempty"`
 
-	// AvailabilityZone is the volume availability zone to create the volume in.
-	// If omitted, the availability zone of the server will be used.
-	// The availability zone must NOT contain spaces otherwise it will lead to volume that belongs
-	// to this availability zone register failure, see kubernetes/cloud-provider-openstack#1379 for
-	// further information.
+	// AvailabilityZone is the volume availability zone to create the volume
+	// in. If not specified, the volume will be created without an explicit
+	// availability zone.
 	// +optional
-	AvailabilityZone string `json:"availabilityZone,omitempty"`
+	AvailabilityZone *VolumeAvailabilityZone `json:"availabilityZone,omitempty"`
+}
+
+// VolumeAZSource specifies where to obtain the availability zone for a volume.
+// +kubebuilder:validation:Enum=Name;Machine
+type VolumeAZSource string
+
+const (
+	VolumeAZFromName    VolumeAZSource = "Name"
+	VolumeAZFromMachine VolumeAZSource = "Machine"
+)
+
+// VolumeAZName is the name of a volume availability zone. It may not contain spaces.
+// +kubebuilder:validation:Pattern:="^[^ ]+$"
+// +kubebuilder:validation:MinLength:=1
+type VolumeAZName string
+
+// VolumeAvailabilityZone specifies the availability zone for a volume.
+// +kubebuilder:validation:XValidation:rule="!has(self.from) || self.from == 'Name' ? has(self.name) : !has(self.name)",message="name is required when from is 'Name' or default"
+type VolumeAvailabilityZone struct {
+	// From specifies where we will obtain the availability zone for the
+	// volume. The options are "Name" and "Machine". If "Name" is specified
+	// then the Name field must also be specified. If "Machine" is specified
+	// the volume will use the value of FailureDomain, if any, from the
+	// associated Machine.
+	// +kubebuilder:default:=Name
+	// +optional
+	From VolumeAZSource `json:"from,omitempty"`
+
+	// Name is the name of a volume availability zone to use. It is required
+	// if From is "Name". The volume availability zone name may not contain
+	// spaces.
+	// +optional
+	Name *VolumeAZName `json:"name,omitempty"`
```

Comment:
I can't put my finger on why, but this feels clunky. I had envisioned the "inherit from compute AZ" behaviour to be either (a) an anti-feature or (b) if actually required something that would be configured either on a deployment- or machine-wide basis. I've no real reason for this though, other than a general dislike of the whole thing, nor do I have an alternate proposal. As such I'm mainly leaving this note here so that I said _something_.

Reply:
This doesn't preclude adding a machine or cluster-wide default override in the future. However, like you I'm not expecting the `from: Machine` behaviour to be widely used. This is also why I made the discriminator (the From field) optional with an explicit default. This means that the anticipated use case goes from:

```yaml
availabilityZone: foo
```
to
```yaml
availabilityZone:
  name: foo
```

which isn't awful, and gives us a big get-out-of-jail-free card to fix it later without breaking backwards compatibility when we understand it better.

Reply:
IMO the situation gets better now, when the default behavior becomes set no AZ at all.